Part 1 of the project (:

In [39]:
#Loading the dataset and turning it into a DataFrame
from sklearn.datasets import fetch_california_housing
import pandas as pd
import numpy as np

housing_data = fetch_california_housing(as_frame=True)
df=housing_data.frame
df

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0      8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1      8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2      7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3      5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4      3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   
...       ...       ...       ...        ...         ...       ...       ...   
20635  1.5603      25.0  5.045455   1.133333       845.0  2.560606     39.48   
20636  2.5568      18.0  6.114035   1.315789       356.0  3.122807     39.49   
20637  1.7000      17.0  5.205543   1.120092      1007.0  2.325635     39.43   
20638  1.8672      18.0  5.329513   1.171920       741.0  2.123209     39.43   
20639  2.3886      16.0  5.254717   1.162264      1387.0  2.616981     39.37   

       Longitude  MedHouseVal  
0        -122.23        4.526  
1        -122.22        3.585  
2        -122.24        3.521  
3        -122.25        3.413  
4        -122.25        3.422  
...          ...          ...  
20635    -121.09        0.781  
20636    -121.21        0.771  
20637    -121.22        0.923  
20638    -121.32        0.847  
20639    -121.24        0.894  

[20640 rows x 9 columns]

In [43]:
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   MedInc       20640 non-null  float64
 1   HouseAge     20640 non-null  float64
 2   AveRooms     20640 non-null  float64
 3   AveBedrms    20640 non-null  float64
 4   Population   20640 non-null  float64
 5   AveOccup     20640 non-null  float64
 6   Latitude     20640 non-null  float64
 7   Longitude    20640 non-null  float64
 8   MedHouseVal  20640 non-null  float64
dtypes: float64(9)
memory usage: 1.4 MB


MedInc      HouseAge      AveRooms     AveBedrms    Population  \
count  20640.000000  20640.000000  20640.000000  20640.000000  20640.000000   
mean       3.870671     28.639486      5.429000      1.096675   1425.476744   
std        1.899822     12.585558      2.474173      0.473911   1132.462122   
min        0.499900      1.000000      0.846154      0.333333      3.000000   
25%        2.563400     18.000000      4.440716      1.006079    787.000000   
50%        3.534800     29.000000      5.229129      1.048780   1166.000000   
75%        4.743250     37.000000      6.052381      1.099526   1725.000000   
max       15.000100     52.000000    141.909091     34.066667  35682.000000   

           AveOccup      Latitude     Longitude   MedHouseVal  
count  20640.000000  20640.000000  20640.000000  20640.000000  
mean       3.070655     35.631861   -119.569704      2.068558  
std       10.386050      2.135952      2.003532      1.153956  
min        0.692308     32.540000   -124.350000      0.149990  
25%        2.429741     33.930000   -121.800000      1.196000  
50%        2.818116     34.260000   -118.490000      1.797000  
75%        3.282261     37.710000   -118.010000      2.647250  
max     1243.333333     41.950000   -114.310000      5.000010

We can notice that dataset have 20640 rows and 9 columnts
realisticly the target feture should be MedHouseVal where we use the rest to pridict it 

In [45]:
#Splitting the data 
from sklearn.model_selection import train_test_split
x=df.drop('MedHouseVal',axis=1)
y=df['MedHouseVal']
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)
print(f"\nTrain size: {x_train.shape[0]}, Test size: {x_test.shape[0]}")


Train size: 16512, Test size: 4128


Part 2 

In [ ]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
def eval_metrics(true,pred):
    rmse = np.sqrt(mean_squared_error(true,pred))
    mae = mean_absolute_error(true,pred)
    r2 = r2_score(true,pred)
    return rmse,mae,r2

In [8]:
#Using linear regression as a base model
from sklearn.linear_model import LinearRegression
base_model = LinearRegression()
base_model.fit(x_train, y_train)
y_pred = base_model.predict(x_test)


In [ ]:
#Evaluate the model


rmse = np.sqrt(mean_squared_error(y_test, y_pred))
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)


print(f"RMSE: {rmse:.2f}")
print(f"MAE: {mae:.2f}")
print(f"R²: {r2:.2f}")


RMSE: 0.75
MAE: 0.53
R²: 0.58


In [16]:
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
ridge_model=Ridge()
lasso_model=Lasso()
ridge_model.fit(x_train,y_train)
lasso_model.fit(x_train,y_train)
y_ridge_pridect = ridge_model.predict(x_test)
y_lasso_pridict = lasso_model.predict(x_test)


In [31]:
ridge_rmse = np.sqrt(mean_squared_error(y_test,y_ridge_pridect))
ridge_mae = mean_absolute_error(y_test,y_ridge_pridect)
ridge_r2 = r2_score(y_test,y_ridge_pridect)
print(f"Ridge RMSE: {ridge_rmse:.2f}")
print(f"Ridge MAE:{ridge_mae:.2f}")
print(f"Ridge R²: {ridge_r2:.2f}")
print("============================")
lasso_rmse = np.sqrt(mean_squared_error(y_test,y_lasso_pridict))
lasso_mae = mean_absolute_error(y_test,y_lasso_pridict)
lasso_r2 = r2_score(y_test,y_lasso_pridict)
print(f"Lasso RMSE: {lasso_rmse:.2f}")
print(f"Lasso MAE:{lasso_mae:.2f}")
print(f"Lasso R²: {lasso_r2:.2f}")

Ridge RMSE: 0.75
Ridge MAE:0.53
Ridge R²: 0.58
Lasso RMSE: 0.97
Lasso MAE:0.76
Lasso R²: 0.28


We can notice that Riddge without any tunning is pretty much identical to the base liner regression model as for the lasso model it's much worse noticing almost having a 1 for RMSE and a very low R^2 value of around 28% only compared to over 58% on the other models .

In [35]:
#Adding polynomial features for all models
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler

model_ridge_poly = make_pipeline(
    PolynomialFeatures(degree=2, include_bias=False),
    StandardScaler(),
    Ridge()
)
model_ridge_poly.fit(x_train, y_train)
y_pred_r_p = model_ridge_poly.predict(x_test)

model_lasso_poly = make_pipeline(
    PolynomialFeatures(degree=2,include_bias=False),
    StandardScaler(),
    Lasso()
)
model_lasso_poly.fit(x_train,y_train)
y_pred_l_p = model_lasso_poly.predict(x_test)

model_poly = make_pipeline(
    PolynomialFeatures(degree=2 , include_bias=False),
    StandardScaler(),
    LinearRegression()
)
model_poly.fit(x_train,y_train)
y_pred_m_p=model_lasso_poly.predict(x_test)

linear_rmse = np.sqrt(mean_squared_error(y_test, y_pred_m_p))
linear_mae = mean_absolute_error(y_test, y_pred_m_p)
linear_r2 = r2_score(y_test, y_pred_m_p)

ridge_rmse = np.sqrt(mean_squared_error(y_test, y_pred_r_p))
ridge_mae = mean_absolute_error(y_test, y_pred_r_p)
ridge_r2 = r2_score(y_test, y_pred_r_p)

lasso_rmse = np.sqrt(mean_squared_error(y_test, y_pred_l_p))
lasso_mae = mean_absolute_error(y_test, y_pred_l_p)
lasso_r2 = r2_score(y_test, y_pred_l_p)


print(f"Linear (Poly) RMSE: {linear_rmse:.2f}, MAE: {linear_mae:.2f}, R²: {linear_r2:.2f}")
print(f"Ridge (Poly) RMSE: {ridge_rmse:.2f}, MAE: {ridge_mae:.2f}, R²: {ridge_r2:.2f}")
print(f"Lasso (Poly) RMSE: {lasso_rmse:.2f}, MAE: {lasso_mae:.2f}, R²: {lasso_r2:.2f}\n")

results_df = pd.DataFrame({
    'Model': ['Linear (Poly)', 'Ridge (Poly)', 'Lasso (Poly)'],
    'RMSE': [linear_rmse, ridge_rmse, lasso_rmse],
    'MAE': [linear_mae, ridge_mae, lasso_mae],
    'R²': [linear_r2, ridge_r2, lasso_r2]
})

print(results_df)


Linear (Poly) RMSE: 1.14, MAE: 0.91, R²: -0.00
Ridge (Poly) RMSE: 0.69, MAE: 0.49, R²: 0.64
Lasso (Poly) RMSE: 1.14, MAE: 0.91, R²: -0.00

           Model      RMSE       MAE        R²
0  Linear (Poly)  1.144856  0.906069 -0.000219
1   Ridge (Poly)  0.687665  0.487971  0.639133
2   Lasso (Poly)  1.144856  0.906069 -0.000219
